sql: add method for setting "external" query plan hints

DrewKimball opened the pull request:
#### sql: add system.plan_hints internal table

This patch adds a new table, `system.plan_hints`, which can be used to
associate "external" plan hints with a query, without having to modify
the query itself. A following commit will add a per-node cache for this
table, similar to the way table stats are handled.

Epic: None

Release note: None

#### sql: add per-node cache for system.plan_hints

This commit adds a per-node cache for the `system.plan_hints` internal
table, similar to the table stats cache. It also adds a builtin function
`crdb_internal.query_id_for_hint` for convenience, so that query IDs
for compatible with the table can be computed from the CLI.

Epic: None

Release note: None

#### sql,hints: add example hint for SET statements

This commit adds an example hint for the `system.plan_hints` table, which
allows the user to specify `SET` statements to run before executing a
query. This has some rough edges and TODOs, but seems to work so far.

Epic: None

Release note: None

DrewKimball committed:
sql: add system.plan_hints internal table

This patch adds a new table, `system.plan_hints`, which can be used to
associate "external" plan hints with a query, without having to modify
the query itself. A following commit will add a per-node cache for this
table, similar to the way table stats are handled.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119120)
<!-- Reviewable:end -->

DrewKimball commented:
I added builtin functions to conveniently add plan hints, but that's probably not a good UI, and possibly also not safe, since it involves mutating the `plan_hints` table. For a real implementation, we'd probably want to figure out something better.

DrewKimball commented:
Example usage:
```
CREATE TABLE ab (a INT, b INT);
CREATE TABLE xy (x INT, y INT);
INSERT INTO ab SELECT t, t*2 FROM generate_series(1, 50);
INSERT INTO xy SELECT t, t*2 FROM generate_series(1, 10000);
ANALYZE ab;
ANALYZE xy;

-- Optimal plan is a hash join with xy on the left.
EXPLAIN SELECT * FROM ab JOIN xy ON a = x;

-- Get the query ID.
SELECT crdb_internal.query_id_for_hint('SELECT * FROM ab JOIN xy ON a = x');

-- Add a plan hint to disable join reordering (including commutation.
SELECT crdb_internal.hint_setting(<id>, 'reorder_joins_limit', '0');

-- The join should be reversed.
EXPLAIN SELECT * FROM ab JOIN xy ON a = x;

-- The modified setting should show up in the env file of a statement bundle.
EXPLAIN ANALYZE SELECT * FROM ab JOIN xy ON a = x;

-- The setting should not be changed after the statement completes.
SHOW reorder_joins_limit;
```

DrewKimball committed:
sql: add per-node cache for system.plan_hints

This commit adds a per-node cache for the `system.plan_hints` internal
table, similar to the table stats cache. It also adds the `PlanHints`
protobuf struct, which will be used to get the bytes representation of
plan hints for storage in the `system.plan_hints` table.

Epic: None

Release note: None

dt commented on pkg/ccl/backupccl/system_schema.go:
```diff
@@ -814,6 +814,10 @@ var systemTableBackupConfiguration = map[string]systemBackupConfiguration{
 	},
+	systemschema.PlanHintsTable.GetName(): {
+		shouldIncludeInClusterBackup: optInToClusterBackup, // No desc ID columns.
+		customRestoreFunc:            defaultSystemTableRestoreFunc,
```
Do we think we'll want to backup/restore plan hints that are applicable to queries that hit a particular table or tables in a particular database when we backup and restore that table or database, even if they're not backing up/restoring the whole cluster?

DrewKimball committed:
hints: add external plan hint for session settings

DrewKimball committed:
sql: add AST external plan hint

This is a reference/prototype implementation of an "AST" external plan
hint, which allows join and index hints to be applied to a matched
query fingerprint. For example. if the fingerprint is:
```
SELECT * FROM ab INNER JOIN xy ON a = x AND b = _
```
The hinted fingerprint could be:
```
SELECT * FROM ab@b_idx INNER HASH JOIN xy ON a = x AND b = _
```

DrewKimball committed:
add walkable ast node interface
